File: October/24/intersect2linkedlist.cpp
// LeetCode 160: Intersection of Two Linked Lists
// Explanation:
// 1. Use two pointers moving through both lists.
// 2. When one reaches the end, redirect to the other’s head.
// 3. They will meet at intersection or both end at null.
// Time Complexity: O(m + n)
// Space Complexity: O(1)

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(nullptr) {}
};

class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        if (!headA || !headB) return nullptr;
        ListNode *a = headA, *b = headB;
        while (a != b) {
            a = a ? a->next : headB;
            b = b ? b->next : headA;
        }
        return a;
    }
};

#include <iostream>
using namespace std;

int main() {
    ListNode* intersect = new ListNode(8);
    intersect->next = new ListNode(4);
    intersect->next->next = new ListNode(5);

    ListNode* headA = new ListNode(4);
    headA->next = new ListNode(1);
    headA->next->next = intersect;

    ListNode* headB = new ListNode(5);
    headB->next = new ListNode(0);
    headB->next->next = new ListNode(1);
    headB->next->next->next = intersect;

    Solution sol;
    ListNode* res = sol.getIntersectionNode(headA, headB);
    if (res) cout << res->val << endl;  // Output: 8
    else cout << "No intersection" << endl;
}

File: 2025/October/24/intersect2linkedlist.py
# LeetCode 160: Intersection of Two Linked Lists
# Explanation:
# 1. Use two pointers `a` and `b` starting at headA and headB.
# 2. When either pointer reaches the end, redirect it to the other list’s head.
# 3. If the lists intersect, pointers will meet at the intersection node.
# 4. If not, both will reach None at the same time.
# Time Complexity: O(m + n)
# Space Complexity: O(1)

class Solution:
    def getIntersectionNode(self, headA, headB):
        if not headA or not headB:
            return None
        a, b = headA, headB
        while a != b:
            a = a.next if a else headB
            b = b.next if b else headA
        return a

# Example setup:
# A: 4 -> 1 \
#              8 -> 4 -> 5
# B:    5 -> 0 -> 1 /
#
# Intersection at node with value 8

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

# Build example lists
intersect = ListNode(8, ListNode(4, ListNode(5)))
headA = ListNode(4, ListNode(1, intersect))
headB = ListNode(5, ListNode(0, ListNode(1, intersect)))

print(Solution().getIntersectionNode(headA, headB).val)  # Output: 8
